### PR TITLE
Add profile retrieval option to retrieve command

### PR DESCRIPTION
**Pull request by kenhuman**
This adds an option to retrieve complete profiles. Additionally, updated CommandViewService with a function to update the labels on TreeView objects. Retrieve operations can be a bit time consuming, so the retrieve command uses this new function to provide some feedback on the status of the operation.

### CODE REVIEW

**Comment by kenhuman** [comment]
I forgot two data types in the profile retrieval code, those are now added. I've also added a new command to open the Flow Builder with the currently open/selected flow.

**Comment by kenhuman** [comment]
Found/fixed a bug introduced in implementation of profile retrieve command, caused commands such as "ForceCode: Refresh from server" to hang. 

Also updated flow builder function to automatically retrieve flow when browser window is closed. Creates a new dependency on package "opn" and requires use of chrome as browser.

File: src/commands/retrieve.ts [hunk 1]
```diff
@@ -170,14 +176,30 @@ export default function retrieve(resource?: vscode.Uri | ToolingTypes) {
...
-            });
-            theStream.on('error', (error) => {
-                reject(error || { message: 'Cannot retrieve more than 10,000 files at a time. Please select "Choose Types..." from the retrieve menu and try to download without Reports selected first.' });
-            });
-            resolve(theStream.stream());
+            var theStream;
+            try {
+                theStream = vscode.window.forceCode.conn.metadata.retrieve({
+                    unpackaged: retrieveTypes,
+                    apiVersion: vscode.window.forceCode.config.apiVersion || constants.API_VERSION,
```
Comment (celador) [inline src/commands/retrieve.ts]: Lets add `|| vscode.workspace.getConfiguration('force')['defaultApiVersion']` here.  It was added in the previous commit

File: src/commands/retrieve.ts [hunk 2]
```diff
@@ -121,17 +122,22 @@ export default function retrieve(resource?: vscode.Uri | ToolingTypes) {
...
                 matchOnDetail: true,
                 placeHolder: 'Retrieve Package',
             };
             return vscode.window.showQuickPick(options, config).then(res => {
-            if (!res) {
-                return Promise.reject();
-            }
-            return res;
-        });
+                if (!res) {
```
Comment (celador) [inline src/commands/retrieve.ts]: fixing indentation should remove this change

File: src/commands/retrieve.ts [hunk 3]
```diff
@@ -276,31 +342,36 @@ export default function retrieve(resource?: vscode.Uri | ToolingTypes) {
...
+                                    id: '',//metadataFileProperties.id,
+                                    lastModifiedDate: '',//metadataFileProperties.lastModifiedDate,
+                                    lastModifiedByName: '',
+                                    lastModifiedById: '',
+                                    type: tType,
+                                    saveTime: true
+                                }
 
-                            newWSMembers.push(wsMem);
+                                newWSMembers.push(wsMem);
```
Comment (celador) [inline src/commands/retrieve.ts]: indentation

File: src/commands/retrieve.ts [hunk 4]
```diff
@@ -276,31 +342,36 @@ export default function retrieve(resource?: vscode.Uri | ToolingTypes) {
...
+                            }
+                            if (name.endsWith('.resource-meta.xml')) {
+                                resBundleNames.push(name);
                             }
                         }
-                        if (name.endsWith('.resource-meta.xml')) {
-                            resBundleNames.push(name);
-                        }
-                        if (name !== 'package.xml' || vscode.window.forceCode.config.overwritePackageXML) {
+                        if ((name !== 'package.xml' || vscode.window.forceCode.config.overwritePackageXML) &&
```
Comment (celador) [inline src/commands/retrieve.ts]: this is a bit hard to follow.  could we extract this test into a function?

File: package.json [hunk 5]
```diff
@@ -494,13 +512,15 @@
     "webpack-dev": "webpack --mode development --watch"
   },
   "dependencies": {
+    "@types/opn": "^5.1.0",
```
Comment (daishi4u) [inline package.json]: Types should be in the dev dependencies

File: package.json [hunk 6]
```diff
@@ -494,13 +512,15 @@
...
     "compressing": "1.4.0",
     "deepmerge": "3.1.0",
     "fs-extra": "7.0.1",
     "globule": "1.2.1",
     "jsforce": "1.8.5",
     "klaw": "3.0.0",
     "mime-types": "2.1.21",
+    "opn": "^5.4.0",
```
Comment (daishi4u) [inline package.json]: I like to go with certain versions instead of using ^ as this has broken Forcecode before

File: src/commands/retrieve.ts [hunk 7]
```diff
@@ -170,14 +176,30 @@ export default function retrieve(resource?: vscode.Uri | ToolingTypes) {
...
+                    unpackaged: retrieveTypes,
+                    apiVersion: vscode.window.forceCode.config.apiVersion || constants.API_VERSION,
+                });
+            } catch(e) {
+                reject(e);
+            }
+            if(theStream) {
+                theStream
+                    .on('progress', (result) => {
+                        CommandViewService.getInstance().updateLabel(task, `Retrieving package (${result.id} - ${result.state})`);
```
Comment (daishi4u) [inline src/commands/retrieve.ts]: You can use commandViewService instead of CommandViewService.getInstance(), it's previously declared as a constant

File: src/models/commands.ts [hunk 8]
```diff
@@ -419,6 +424,60 @@ export const fcCommands: FCCommand[] = [
...
+										break;
+									case 'linux':
+										options.app = ['google-chrome', `--user-data-dir=${randomDir}`];
+										break;
+								}
+
+								dxService.getOrgPageUrl(`/builder_platform_interaction/flowBuilder.app?isFromAloha=false&flowId=${res.records[0].Id}`).then(response => {
+									CommandViewService.getInstance().updateLabel(task, `Editing ${label}`);
+									opn(response.url, options).then(child => {
+										resolve(commands.retrieve(context, task));
```
Comment (daishi4u) [inline src/models/commands.ts]: Will this actually wait until the user is done using Flow Builder before it resolves? It's a really neat idea, but I don't like the fact that it only works with Chrome. Maybe a section in the readme for the new option would be a good idea, and to let users know it will only work with chrome.

**Comment by daishi4u** [comment]
Closing due to inactivity
